asteroid/complex_nn.py

```diff
@@ -0,0 +1,172 @@
+    the real and imaginary components independently.
+
+    :Return:
+        cf(x), complex version of `f`: A function that applies `f` to the real and
+        imaginary components of `x` and returns the result as PyTorch complex tensor.
+    """
+
+    @functools.wraps(f)
+    def cf(x):
+        return as_torch_complex((f(x.real), f(x.imag)))
+
+    # functools.wraps keeps the original name of `f`, which might be confusing,
+    # since we are creating a new function that behaves differently.
+    # Both __name__ and __qualname__ are used by printing code.
+    cf.__name__ == f"{f.__name__} (complex)"
+    cf.__qualname__ == f"{f.__qualname__} (complex)"
+    return cf
+
+
+class OnComponents(nn.Module):
+    """Like `on_components`, but for stateful modules.
+
+    Args:
+        module_cls (callable): A class or function that returns a Torch module/functional.
+            Called 2x with *args, **kwargs, to construct the real and imaginary component modules.
+    """
+
+    def __init__(self, module_cls, *args, **kwargs):
+        super().__init__()
+        self.re_module = module_cls(*args, **kwargs)
+        self.im_module = module_cls(*args, **kwargs)
+
+    def forward(self, x):
+        return as_torch_complex((self.re_module(x.real), self.im_module(x.imag)))
+
+
+class ComplexMultiplicationWrapper(nn.Module):
+    """TODO: Name
+
+    Args:
+        module_cls (callable): A class or function that returns a Torch module/functional.
+            Called 2x with *args, **kwargs, to construct the real and imaginary component modules.
+    """
+
+    def __init__(self, module_cls, *args, **kwargs):
+        super().__init__()
+        self.re_module = module_cls(*args, **kwargs)
+        self.im_module = module_cls(*args, **kwargs)
+
+    def forward(self, x: ComplexTensor) -> ComplexTensor:
+        return as_torch_complex(
+            (
+                self.re_module(x.real) - self.im_module(x.imag),
+                self.re_module(x.imag) + self.im_module(x.real),
+            )
+        )
+
+
+ComplexConv2d = functools.partial(ComplexMultiplicationWrapper, nn.Conv2d)
+ComplexConvTranspose2d = functools.partial(ComplexMultiplicationWrapper, nn.ConvTranspose2d)
+
+
+class BoundComplexMask(nn.Module):
+    """Module version of `bound_complex_mask`"""
+
+    def __init__(self, bound_type):
+        super().__init__()
+        self.bound_type = bound_type
+
+    def forward(self, mask: ComplexTensor):
+        return bound_complex_mask(mask, self.bound_type)
+
+
+def bound_complex_mask(mask: ComplexTensor, bound_type="tanh"):
+    r"""Bound a complex mask, as proposed in [1], section 3.2.
+
+    Valid bound types, for a complex mask $M = |M| \cdot e^{i \varphi(M)}$:
+
+    - Unbounded ("UBD"): $M_{\mathrm{UBD}} = M$
+    - Sigmoid ("BDSS"): $M_{\mathrm{BDSS}} = \sigma(|M|) e^{i \sigma(\varphi(M))}$
+    - Tanh ("BDT"): $M_{\mathrm{BDT}} = \mathrm{tanh}(|M|) e^{i \varphi(M)}$
```

Comment:
Added formulas, is this what you wanted? sqrt(2) is a possible magnitude value for BDSS.